maro/simulator/scenarios/finance/business_engine.py

```diff
@@ -0,0 +1,653 @@
+        # stock_num = len(self._stocks)
+
+        # self._distance_adj = distance_adj.reshape(station_num, station_num)
+
+        # Add wrapper to it to make it easy to use, with this we can get value by:
+        # 1. self._trips_adj[x, y].
+        # 2. self._trips_adj.get_row(0).
+        # 3. self._trips_adj.get_column(0).
+        # self._trips_adj = MatrixAttributeAccessor(self._matrices_node, "trips_adj", station_num, station_num)
+        pass
+
+    def _init_frame(self):
+        self._frame = build_frame(len(self._conf["stocks"]), self.calc_max_snapshots())
+        self._snapshots = self._frame.snapshots
+        self._account = self._frame.account[0]
+        self._account.set_init_state(
+            init_money=self._conf['account']['money']
+        )
+
+    def _register_events(self):
+        # Register our own events and their callback handlers.
+        self._event_buffer.register_event_handler(DECISION_EVENT, self._on_action_recieved)
+
+    def is_market_closed(self):
+        # TODO: Implement
+        return True
+
+    def _on_action_recieved(self, evt: Event):
+        actions = evt.payload
+        if actions is None:
+            return
+
+        for action in actions:
+            if action is None:
+                pass
+            else:
+                if isinstance(action, CancelOrder):
+                    self.cancel_order(action, evt.tick)
+                else:
+                    results: list = self.take_action(
+                        action, evt.tick
+                    )
+                    for result in results:
+                        self._account.take_trade(
+                            action, result, self._stocks
+                        )
+
+                if action.state != ActionState.pending and action.id not in self._finished_action:
+                    self._finished_action[action.id] = action
+        # append panding orders still not triggered in the available life time
+        if evt.tick in self._pending_orders:
+            for action in self._pending_orders[evt.tick]:
+                self._event_buffer.gen_atom_event(evt.tick, DecisionEvent, action)
+            self._pending_orders[evt.tick].clear()
+
+    def take_action(self, action: Action, tick: int) -> TradeResult:
+        # 1. can trade -> bool
+        # 2. return (stock, sell/busy, stock_price, number, tax)
+        # 3. update stock.account_hold_num
+        rets = []
+        if action.state == ActionState.pending:
+            # not canceled
+            stock = self._stocks[action.item]
+            is_trigger = action.is_trigger(stock.opening_price, stock.trade_volume)
+            if is_trigger:
+                print(f"  <<>>  executing {action}")
+                if not self._allow_split:
+                    rets = self.trade(action)
+                else:
+                    # TODO: Split trade should be implemented
+                    rets = self.split_trade(action)
+
+                if len(rets) > 0:
```

Comment:
can we return ``bool, List[TradeResult]`` in ``self.trade()`` and ``self.split_trade()``? So we can directly use the boolean to adjust success or fail

Reply:
ok